src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,368 @@
+  [ARBITRUM]: "0x70d95587d40A2caf56bd97485aB3Eec10Bee6336", // ETH/USD
+  [AVALANCHE]: "0xB7e69749E3d2EDd90ea59A4932EFEa2D41E245d7", // ETH/USD
+  [AVALANCHE_FUJI]: "0xbf338a6C595f06B7Cfff2FA8c958d49201466374", // ETH/USD
+};
+
+type RpcProbe = {
+  isSuccess: boolean;
+  responseTime: number;
+  blockNumber: number | null;
+  timestamp: Date;
+};
+
+type ProviderData = {
+  url: string;
+  provider: Provider;
+  probeData: RpcProbe[];
+};
+
+type RpcTrackerState = {
+  [chainId: number]: {
+    lastUsage: Date;
+    currentBestProviderUrl: string;
+    providers: {
+      [providerUrl: string]: ProviderData;
+    };
+  };
+};
+
+let trackingIntervalId: NodeJS.Timeout | null = null;
+let trackerState: RpcTrackerState | null = null;
+
+function initRpcTracking() {
+  trackerState = initTrackerState();
+
+  if (trackingIntervalId) {
+    clearInterval(trackingIntervalId);
+  }
+  measureRpcData();
+  trackingIntervalId = setInterval(measureRpcData, PROBE_INTERVAL);
+}
+
+function measureRpcData() {
+  if (!trackerState) {
+    throw new Error("RPC tracker state is not initialized");
+  }
+
+  entries(trackerState).forEach(async ([chainIdRaw, chainTracker]) => {
+    const chainId = Number(chainIdRaw);
+    const providers = Object.values(chainTracker.providers);
+
+    if (
+      differenceInMilliseconds(Date.now(), chainTracker.lastUsage) > DISABLE_UNUSED_TRACKING_TIMEOUT ||
+      providers.length < 2
+    ) {
+      return;
+    }
+
+    const probes = providers.map(async (providerInfo) => {
+      providerInfo.probeData = providerInfo.probeData.filter((probe) => {
+        return Date.now() - probe.timestamp.getTime() < PROBE_SLIDING_WINDOW;
+      });
+
+      return probeRpc(chainId, providerInfo.provider, providerInfo.url).then((probeResult) => {
+        providerInfo.probeData.push(probeResult);
+
+        return probeResult;
+      });
+    });
+
+    const probeResults = await Promise.all(probes);
+    const successProbeResults = probeResults.filter((probe) => probe.isSuccess);
+
+    if (!successProbeResults.length) {
+      const fallbackUrl = getBestRpc(chainId).fallback;
+
+      if (fallbackUrl) {
+        setCurrentProvider(chainId, fallbackUrl);
+      }
+
+      return;
+    }
+
+    const probeResultsByBlockNumber = orderBy(successProbeResults, ["blockNumber"], ["desc"]);
+    let bestBlockNumberProbe = probeResultsByBlockNumber[0];
+
+    const probeStats = probeResultsByBlockNumber.map((probe, i, arr) => {
+      let isValid = probe.isSuccess;
+
+      const bestBlockNumber = bestBlockNumberProbe?.blockNumber;
+      const currProbeBlockNumber = probe.blockNumber;
+      const nextProbeBlockNumber = arr[i + 1]?.blockNumber;
+
+      // Rare case when RPC returned a block number from the far future
+      if (
+        i === 0 &&
+        currProbeBlockNumber &&
+        nextProbeBlockNumber &&
+        currProbeBlockNumber - nextProbeBlockNumber > BLOCK_FROM_FUTURE_THRESHOLD
+      ) {
+        bestBlockNumberProbe = probe;
```

Comment:
I think you want `bestBlockNumberProbe = arr[i + 1]` here?

Reply:
Indeed, you are right